src/Cache/CachingResult.php

```diff
@@ -43,18 +43,23 @@ class CachingResult implements Result
     /** @var array<int,array<string,mixed>>|null */
     private $data;
 
+    /** @var array<string, mixed> */
+    private $fetchedData;
+
     /**
-     * @param string $cacheKey
-     * @param string $realKey
-     * @param int    $lifetime
+     * @param string               $cacheKey
+     * @param string               $realKey
+     * @param int                  $lifetime
+     * @param array<string, mixed> $fetchedData
      */
-    public function __construct(Result $result, Cache $cache, $cacheKey, $realKey, $lifetime)
+    public function __construct(Result $result, Cache $cache, $cacheKey, $realKey, $lifetime, $fetchedData)
```

Comment:
See https://github.com/doctrine/dbal/pull/4169#discussion_r460100582.

Reply:
`$fetchedData` - is data, fetched from the result cache by `$cacheKey` key. It is a hash of `$realKey` to serialized `$result`.
One result cache data may store more than 1 result caches, that are distinguished by realKey. 

`CachingResult` is updating one of these keys with the underlying `Result`. 

In order to keep all other keys, it should have the full data, stored in cache. It is fetching it now
https://github.com/doctrine/dbal/blob/824d2eb0af416006f8a5317d08882ea204f608f9/src/Cache/CachingResult.php#L173
and than appends its key to fetched array
https://github.com/doctrine/dbal/blob/824d2eb0af416006f8a5317d08882ea204f608f9/src/Cache/CachingResult.php#L179

However, this array was fetched from cache before 
https://github.com/doctrine/dbal/blob/824d2eb0af416006f8a5317d08882ea204f608f9/src/Connection.php#L1006
and might be just passed to the constructor of `CachingResult`

This PR does exactly this - removes 2nd fetching from CachingResult and passes fetched data from Connection.

Reply:
@k0ka thank you for the details. I'm still trying to get to the bottom of this. The last time I touched this layer, I described the responsibility of this class as following (https://github.com/doctrine/dbal/pull/4048#issuecomment-640071573):

> This class is responsible only for reading from the underlying statement and storing the result in the cache if there was no hit.

Am I wrong, or, otherwise, why does the class try to fetch from the cache if there knowingly was no hit?

Reply:
In any event, as I asked earlier in https://github.com/doctrine/dbal/pull/4169#pullrequestreview-454380957, this change will require a unit test that will also help find the most optimal solution for this problem. Right now, it looks dirty that besides the `Result` that contains the fetched data, the object needs to receive the data fetched from that result.

Reply:
> 
> Am I wrong, or, otherwise, why does the class try to fetch from the cache if there knowingly was no hit?

Because one cache key may contain more than one result.

Reply:
> it looks dirty that besides the `Result` that contains the fetched data, the object needs to receive the data fetched from that result.

I have added unit and functional test to show you how it works now.

Reply:
> I have added unit and functional test to show you how it works now.

Thanks for the test. I understand the logic better now.

> Because one cache key may contain more than one result.

Do you know why we need to store more than one result in a cache entry? I.e. why not store each key individually and use the cache key as a sort of namespace for hashing the keys?

Reply:
> Do you know why we need to store more than one result in a cache entry?

I don't know exactly, but I think it's done to be able to easily clear cache entries. You don't have to know exact sql and all parameters to purge its result.

Reply:
```suggestion
    public function __construct(Result $result, Cache $cache, $cacheKey, $realKey, $lifetime, array $fetchedData)
```